# Welcome to todays R-excercises - Managing data frames
Today you will be merging and playing around with data frames and learn how to read a CSV table through a case of demographic income and trends. 
Note: Data for this section was sourced from data.worldbank.org

# Case

Think this: You are a data scientist hired by 'The unethical bank of the Caymen islands' who needs to give the bank of the current relation between glboal demography and income. They also want to get some insights into the relationship between income and internet usage globally. 


## Data 
We've got 5 columns. Country name, country code, birth rate, internet usage and income group. 


# Importing data into our R workflow/ jupyter notebook
Firstly we will import the data into our workflow. We will demonstrate two common ways of doing this. This part is crucial to understand so please refer to this file for future work in case your files won't read. When we load a file (CSV or EXCEL) R automatically makes a DATA FRAME for it that you can work on! Make sure the data file is on the root folder
## importing data from a local directory (only in R studio - not in notebook) 
stats &lt;- read.csv(file.choose()) 


## importing data from our workflow (prefered and use this for now) 

In [1]:
suppressWarnings(install.packages("ggplot2"))
library(ggplot2)

package 'ggplot2' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\MaZeeT\AppData\Local\Temp\Rtmp0Q4cXn\downloaded_packages


Warning message:
"package 'ggplot2' was built under R version 3.6.3"

ERROR: Error: package or namespace load failed for 'ggplot2' in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):
 namespace 'rlang' 0.3.4 is already loaded, but >= 0.4.0 is required


In [2]:
data <- read.csv("P2-Demographic-Data.csv")

We assigned the name data to our data frame and can now explore the data with the command:

In [4]:
data

Country.Name                   Country.Code Birth.rate Internet.users
1   Aruba                          ABW          10.244     78.90000      
2   Afghanistan                    AFG          35.253      5.90000      
3   Angola                         AGO          45.985     19.10000      
4   Albania                        ALB          12.877     57.20000      
5   United Arab Emirates           ARE          11.044     88.00000      
6   Argentina                      ARG          17.716     59.90000      
7   Armenia                        ARM          13.308     41.90000      
8   Antigua and Barbuda            ATG          16.447     63.40000      
9   Australia                      AUS          13.200     83.00000      
10  Austria                        AUT           9.400     80.61880      
11  Azerbaijan                     AZE          18.300     58.70000      
12  Burundi                        BDI          44.151      1.30000      
13  Belgium                        BEL          11.200     82.17020      
14  Benin                          BEN          36.440      4.90000      
15  Burkina Faso                   BFA          40.551      9.10000      
16  Bangladesh                     BGD          20.142      6.63000      
17  Bulgaria                       BGR           9.200     53.06150      
18  Bahrain                        BHR          15.040     90.00004      
19  Bahamas, The                   BHS          15.339     72.00000      
20  Bosnia and Herzegovina         BIH           9.062     57.79000      
21  Belarus                        BLR          12.500     54.17000      
22  Belize                         BLZ          23.092     33.60000      
23  Bermuda                        BMU          10.400     95.30000      
24  Bolivia                        BOL          24.236     36.94000      
25  Brazil                         BRA          14.931     51.04000      
26  Barbados                       BRB          12.188     73.00000      
27  Brunei Darussalam              BRN          16.405     64.50000      
28  Bhutan                         BTN          18.134     29.90000      
29  Botswana                       BWA          25.267     15.00000      
30  Central African Republic       CAF          34.076      3.50000      
... ...                            ...          ...        ...           
166 Seychelles                     SYC          18.600     50.40         
167 Syrian Arab Republic           SYR          24.043     26.20         
168 Chad                           TCD          45.745      2.30         
169 Togo                           TGO          36.080      4.50         
170 Thailand                       THA          11.041     28.94         
171 Tajikistan                     TJK          30.792     16.00         
172 Turkmenistan                   TKM          21.322      9.60         
173 Timor-Leste                    TLS          35.755      1.10         
174 Tonga                          TON          25.409     35.00         
175 Trinidad and Tobago            TTO          14.590     63.80         
176 Tunisia                        TUN          19.800     43.80         
177 Turkey                         TUR          16.836     46.25         
178 Tanzania                       TZA          39.518      4.40         
179 Uganda                         UGA          43.474     16.20         
180 Ukraine                        UKR          11.100     41.00         
181 Uruguay                        URY          14.374     57.69         
182 United States                  USA          12.500     84.20         
183 Uzbekistan                     UZB          22.500     38.20         
184 St. Vincent and the Grenadines VCT          16.306     52.00         
185 Venezuela, RB                  VEN          19.842     54.90         
186 Virgin Islands (U.S.)          VIR          10.700     45.30         
187 Vietnam                        VNM          15.537     43.90         
188 Vanu

## Exploring the data set
Let us try and explore the data set now. We have a large data set with alot of rows. It is a bit difficult to look at it this way. let us therefor explore some functions 

In [5]:
stats <- read.csv("P2-Demographic-Data.csv")

In [6]:
nrow(stats)

[1] 195

You see now that we have 195 rows. Let's look at our number of columns.

In [7]:
ncol(stats)

[1] 5

Cool. let's look at two other functions. What are the top 6 rows for example? you might need this to look at a sum-up of your dataset as you have a quite large dataset. We now see a sample 

In [8]:
head(stats)

Country.Name         Country.Code Birth.rate Internet.users
1 Aruba                ABW          10.244     78.9          
2 Afghanistan          AFG          35.253      5.9          
3 Angola               AGO          45.985     19.1          
4 Albania              ALB          12.877     57.2          
5 United Arab Emirates ARE          11.044     88.0          
6 Argentina            ARG          17.716     59.9          
  Income.Group       
1 High income        
2 Low income         
3 Upper middle income
4 Upper middle income
5 High income        
6 High income

Let's look at the bottom 22 rows and get a different sample of the data. Do you notice the difference in the code and the output? why did we get 22 rows? 

In [9]:
tail(data, n=22) 

Country.Name                   Country.Code Birth.rate Internet.users
174 Tonga                          TON          25.409     35.00         
175 Trinidad and Tobago            TTO          14.590     63.80         
176 Tunisia                        TUN          19.800     43.80         
177 Turkey                         TUR          16.836     46.25         
178 Tanzania                       TZA          39.518      4.40         
179 Uganda                         UGA          43.474     16.20         
180 Ukraine                        UKR          11.100     41.00         
181 Uruguay                        URY          14.374     57.69         
182 United States                  USA          12.500     84.20         
183 Uzbekistan                     UZB          22.500     38.20         
184 St. Vincent and the Grenadines VCT          16.306     52.00         
185 Venezuela, RB                  VEN          19.842     54.90         
186 Virgin Islands (U.S.)          VIR          10.700     45.30         
187 Vietnam                        VNM          15.537     43.90         
188 Vanuatu                        VUT          26.739     11.30         
189 West Bank and Gaza             PSE          30.394     46.60         
190 Samoa                          WSM          26.172     15.30         
191 Yemen, Rep.                    YEM          32.947     20.00         
192 South Africa                   ZAF          20.850     46.50         
193 Congo, Dem. Rep.               COD          42.394      2.20         
194 Zambia                         ZMB          40.471     15.40         
195 Zimbabwe                       ZWE          35.715     18.50         
    Income.Group       
174 Upper middle income
175 High income        
176 Upper middle income
177 Upper middle income
178 Low income         
179 Low income         
180 Lower middle income
181 High income        
182 High income        
183 Lower middle income
184 Upper middle income
185 High income        
186 High income        
187 Lower middle income
188 Lower middle income
189 Lower middle income
190 Lower middle income
191 Lower middle income
192 Upper middle income
193 Low income         
194 Lower middle income
195 Low income

Let's check out the structure of the data

In [10]:
str(data) 

'data.frame':	195 obs. of  5 variables:
 $ Country.Name  : Factor w/ 195 levels "Afghanistan",..: 8 1 4 2 183 6 7 5 9 10 ...
 $ Country.Code  : Factor w/ 195 levels "ABW","AFG","AGO",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ Birth.rate    : num  10.2 35.3 46 12.9 11 ...
 $ Internet.users: num  78.9 5.9 19.1 57.2 88 ...
 $ Income.Group  : Factor w/ 4 levels "High income",..: 1 2 4 4 1 1 3 1 1 1 ...


This gives us a quick briefing of categories and variables.. We see the number of rows and columns. A factor w/ 4 levels for example means that there is a variable that is being interpreted as a categorical variable.. so 4 types of income groups. Go back and check the data. is this true? It also assumes that the country name and the country code are factors even though we have 195 levels. Does this fit with the number of countries in our data set? please check it! 
Let's check out another explorative function

In [11]:
summary(data)

              Country.Name  Country.Code   Birth.rate    Internet.users 
 Afghanistan        :  1   ABW    :  1   Min.   : 7.90   Min.   : 0.90  
 Albania            :  1   AFG    :  1   1st Qu.:12.12   1st Qu.:14.52  
 Algeria            :  1   AGO    :  1   Median :19.68   Median :41.00  
 Angola             :  1   ALB    :  1   Mean   :21.47   Mean   :42.08  
 Antigua and Barbuda:  1   ARE    :  1   3rd Qu.:29.76   3rd Qu.:66.22  
 Argentina          :  1   ARG    :  1   Max.   :49.66   Max.   :96.55  
 (Other)            :189   (Other):189                                  
              Income.Group
 High income        :67   
 Low income         :30   
 Lower middle income:50   
 Upper middle income:48   
                          
                          
                          

What do you see here? If you want to read more about this function go to : https://www.r-bloggers.com/r-tutorial-series-summary-and-descriptive-statistics/
Congratulations with your basic exploration of your dataset! let's move on. 

## The famous $ sign


How could you select and extract the birthdata for for example the country Angola? 

In [12]:
data[3,3]

[1] 45.985

Brithrate of Angola should be 45,985. Please check if that is true? we can also try the same but with the name of the column header

In [13]:
data[3,"Birth.rate"]

[1] 45.985

the $-sign is another way of accessing your data without remembering the names of the rows and columns. Let's try it out (might only work in R studio)

In [14]:
data$Internet.users

[1] 78.90000  5.90000 19.10000 57.20000 88.00000 59.90000 41.90000 63.40000
  [9] 83.00000 80.61880 58.70000  1.30000 82.17020  4.90000  9.10000  6.63000
 [17] 53.06150 90.00004 72.00000 57.79000 54.17000 33.60000 95.30000 36.94000
 [25] 51.04000 73.00000 64.50000 29.90000 15.00000  3.50000 85.80000 86.34000
 [33] 66.50000 45.80000  8.40000  6.40000  6.60000 51.70000  6.50000 37.50000
 [41] 45.96000 27.93000 74.10000 65.45480 74.11040 84.17000  9.50000 94.62970
 [49] 45.90000 16.50000 40.35368 29.40000  0.90000 71.63500 79.40000  1.90000
 [57] 91.51440 37.10000 81.91980 27.80000  9.20000 89.84410 43.30000 12.30000
 [65]  1.60000 14.00000  3.10000 16.40000 59.86630 35.00000 65.80000 19.70000
 [73] 65.40000 35.00000 74.20000 17.80000 66.74760 10.60000 72.64390 14.94000
 [81] 15.10000 78.24770 29.95000  9.20000 96.54680 70.80000 58.45930 37.10000
 [89] 41.00000 89.71000 54.00000 39.00000 23.00000  6.80000 11.50000 84.77000
 [97] 75.46000 12.50000 70.50000  3.20000 16.50000 46.20000 93.80000 21.90000
[105]  5.00000 68.45290 93.77650 75.23440 65.80000 56.00000 45.00000  3.00000
[113] 44.10000 43.46000 65.24000  3.50000 68.91380  1.60000 60.31000 20.00000
[121]  5.40000  6.20000 39.00000  5.05000 66.97000 13.90000 66.00000  1.70000
[129] 38.00000 15.50000 93.95640 95.05340 13.30000 82.78000 66.45000 10.90000
[137] 44.03000 39.20000 37.00000  6.50000 62.84920 73.90000 62.09560 36.90000
[145] 56.80000 85.30000 49.76450 67.97000  9.00000 60.50000 22.70000 13.10000
[153] 81.00000  8.00000  1.70000 23.10930  1.50000 51.50000 14.10000 23.00000
[161] 37.40000 77.88260 72.67560 94.78360 24.70000 50.40000 26.20000  2.30000
[169]  4.50000 28.94000 16.00000  9.60000  1.10000 35.00000 63.80000 43.80000
[177] 46.25000  4.40000 16.20000 41.00000 57.69000 84.20000 38.20000 52.00000
[185] 54.90000 45.30000 43.90000 11.30000 46.60000 15.30000 20.00000 46.50000
[193]  2.20000 15.40000 18.50000

we can now see the whole vector or column in the dataset. We actually just extracted it. Try this with other variables ! 
Let's try to extract and exlore the levels of 'income group' 

In [15]:
levels(data$Income.Group)

[1] "High income"         "Low income"          "Lower middle income"
[4] "Upper middle income"

We see all four variables here. 

## Basic operations with data frames 

Now the $-sign is going to come in handy

In [16]:
data[1:10,] #subsetting row 1 - 10 

Country.Name         Country.Code Birth.rate Internet.users
1  Aruba                ABW          10.244     78.9000       
2  Afghanistan          AFG          35.253      5.9000       
3  Angola               AGO          45.985     19.1000       
4  Albania              ALB          12.877     57.2000       
5  United Arab Emirates ARE          11.044     88.0000       
6  Argentina            ARG          17.716     59.9000       
7  Armenia              ARM          13.308     41.9000       
8  Antigua and Barbuda  ATG          16.447     63.4000       
9  Australia            AUS          13.200     83.0000       
10 Austria              AUT           9.400     80.6188       
   Income.Group       
1  High income        
2  Low income         
3  Upper middle income
4  Upper middle income
5  High income        
6  High income        
7  Lower middle income
8  High income        
9  High income        
10 High income

Now lets try and look at two specific rows 

In [17]:
data[c(4,100),]

Country.Name Country.Code Birth.rate Internet.users Income.Group       
4   Albania      ALB          12.877     57.2           Upper middle income
100 Liberia      LBR          35.521      3.2           Low income

The [] gives us the opportunity to extract speficic information from our table. Let's try and extract all data from column 1 and make R give us a vector for this. 

In [18]:
data[,1]

[1] Aruba                          Afghanistan                   
  [3] Angola                         Albania                       
  [5] United Arab Emirates           Argentina                     
  [7] Armenia                        Antigua and Barbuda           
  [9] Australia                      Austria                       
 [11] Azerbaijan                     Burundi                       
 [13] Belgium                        Benin                         
 [15] Burkina Faso                   Bangladesh                    
 [17] Bulgaria                       Bahrain                       
 [19] Bahamas, The                   Bosnia and Herzegovina        
 [21] Belarus                        Belize                        
 [23] Bermuda                        Bolivia                       
 [25] Brazil                         Barbados                      
 [27] Brunei Darussalam              Bhutan                        
 [29] Botswana                       Central African Republic      
 [31] Canada                         Switzerland                   
 [33] Chile                          China                         
 [35] Cote d'Ivoire                  Cameroon                      
 [37] Congo, Rep.                    Colombia                      
 [39] Comoros                        Cabo Verde                    
 [41] Costa Rica                     Cuba                          
 [43] Cayman Islands                 Cyprus                        
 [45] Czech Republic                 Germany                       
 [47] Djibouti                       Denmark                       
 [49] Dominican Republic             Algeria                       
 [51] Ecuador                        Egypt, Arab Rep.              
 [53] Eritrea                        Spain                         
 [55] Estonia                        Ethiopia                      
 [57] Finland                        Fiji                          
 [59] France                         Micronesia, Fed. Sts.         
 [61] Gabon                          United Kingdom                
 [63] Georgia                        Ghana                         
 [65] Guinea                         Gambia, The                   
 [67] Guinea-Bissau                  Equatorial Guinea             
 [69] Greece                         Grenada                       
 [71] Greenland                      Guatemala                     
 [73] Guam                           Guyana                        
 [75] Hong Kong SAR, China           Honduras                      
 [77] Croatia                        Haiti                         
 [79] Hungary                        Indonesia                     
 [81] India                          Ireland                       
 [83] Iran, Islamic Rep.             Iraq                          
 [85] Iceland                        Israel                        
 [87] Italy                          Jamaica                       
 [89] Jordan                         Japan                         
 [91] Kazakhstan                     Kenya                         
 [93] Kyrgyz Republic                Cambodia                      
 [95] Kiribati                       Korea, Rep.                   
 [97] Kuwait                         Lao PDR                       
 [99] Lebanon                        Liberia                       
[101] Libya                          St. Lucia                     
[103] Liechtenstein                  Sri Lanka                     
[105] Lesotho                        Lithuania                     
[107] Luxembourg                     Latvia                        
[109] Macao SAR, China               Morocco                       
[111] Moldova                        Madagascar                    
[113] Maldives                       Mexico                        
[115] Macedonia, FYR                 Mali                          
[117] Malta                          Myanmar              

Is this now a dataframee? let's test! 

In [19]:
is.data.frame(data[,1])

[1] FALSE

Let's try and turn it into one

In [20]:
data[,1,drop=F]

Country.Name                  
1   Aruba                         
2   Afghanistan                   
3   Angola                        
4   Albania                       
5   United Arab Emirates          
6   Argentina                     
7   Armenia                       
8   Antigua and Barbuda           
9   Australia                     
10  Austria                       
11  Azerbaijan                    
12  Burundi                       
13  Belgium                       
14  Benin                         
15  Burkina Faso                  
16  Bangladesh                    
17  Bulgaria                      
18  Bahrain                       
19  Bahamas, The                  
20  Bosnia and Herzegovina        
21  Belarus                       
22  Belize                        
23  Bermuda                       
24  Bolivia                       
25  Brazil                        
26  Barbados                      
27  Brunei Darussalam             
28  Bhutan                        
29  Botswana                      
30  Central African Republic      
... ...                           
166 Seychelles                    
167 Syrian Arab Republic          
168 Chad                          
169 Togo                          
170 Thailand                      
171 Tajikistan                    
172 Turkmenistan                  
173 Timor-Leste                   
174 Tonga                         
175 Trinidad and Tobago           
176 Tunisia                       
177 Turkey                        
178 Tanzania                      
179 Uganda                        
180 Ukraine                       
181 Uruguay                       
182 United States                 
183 Uzbekistan                    
184 St. Vincent and the Grenadines
185 Venezuela, RB                 
186 Virgin Islands (U.S.)         
187 Vietnam                       
188 Vanuatu                       
189 West Bank and Gaza            
190 Samoa                         
191 Yemen, Rep.                   
192 South Africa                  
193 Congo, Dem. Rep.              
194 Zambia                        
195 Zimbabwe

Let us test if it is now? 

In [21]:
is.data.frame(data[,1,drop=F])

[1] TRUE

You can also multiply two variables. For example: 

In [22]:
data$Birth.rate * data$Internet.users

[1]  808.2516  207.9927  878.3135  736.5644  971.8720 1061.1884  557.6052
  [8] 1042.7398 1095.6000  757.8167 1074.2100   57.3963  920.3062  178.5560
 [15]  369.0141  133.5415  488.1658 1353.6006 1104.4080  523.6930  677.1250
 [22]  775.8912  991.1200  895.2778  762.0782  889.7240 1058.1225  542.2066
 [29]  379.0050  119.2660  935.2200  880.6680  890.1025  554.1800  313.4880
 [36]  238.3104  244.2726  831.1292  223.1190  810.9375  690.4111  290.4720
 [43]  926.2500  748.5411  755.9261  715.4450  242.1170  946.2970  972.9882
 [50]  408.1770  850.2521  824.1408   31.3200  651.8785  817.8200   62.5575
 [57]  979.2041  759.1773 1007.6135  653.6058  281.1060 1096.0980  577.2756
 [64]  407.5113   59.7392  595.3500  116.2593  579.9368  508.8636  676.6900
 [71]  954.1000  541.0605 1137.2406  660.9750  586.1800  384.3554  627.4274
 [78]  268.6570  668.3239  303.2372  306.3941 1173.7155  536.1050  286.0556
 [85] 1293.7271 1508.0400  496.9040  502.3340 1108.8860  735.6220 1227.4200
 [92] 1372.5660  625.6000  166.3416  334.0060  729.0220 1552.5895  338.1375
 [99]  946.5330  113.6672  353.5125  712.8660  862.9600  391.1997  143.6900
[106]  691.3743 1059.6744  767.3909  740.6448 1177.2880  546.3450  104.0580
[113]  945.8127  830.2598  732.1233  154.4830  654.6811   28.9904  700.5610
[120]  485.5000  214.4070  209.5662  425.1000  199.2680 1125.4308  416.1243
[127] 1122.0000   84.4237 1521.7100  322.2140  958.3553 1102.6194  278.2759
[134] 1086.0736 1356.8426  322.4438  866.5104  791.7616  880.2300  187.8435
[141]  603.3523  798.1200  490.5552  796.5972  931.1224 1018.4820  437.9276
[148]  897.2040  294.2010 1244.8480  759.9279  504.7823  753.3000  244.6240
[155]   62.4393  403.8581   65.8365  473.8000  523.4766  794.3510  690.2170
[162]  786.6143  741.2911 1118.4465  743.2971  937.4400  629.9266  105.2135
[169]  162.3600  319.5265  492.6720  204.6912   39.3305  889.3150  930.8420
[176]  867.2400  778.6650  173.8792  704.2788  455.1000  829.2361 1052.5000
[183]  859.5000  847.9120 1089.3258  484.7100  682.0743  302.1507 1416.3604
[190]  400.4316  658.9400  969.5250   93.2668  623.2534  660.7275

Try and add more variable combinations yourself.

Data frames is the central data types we will be using again and again! Usually, however, we will not "create" them ourselves, but get them using packages for getting data into R. Here is another example of a data frame and how to subset it :

In [23]:
data$MupltiplyCalc <- data$Birth.rate * data$Internet.users
head(data)

Country.Name         Country.Code Birth.rate Internet.users
1 Aruba                ABW          10.244     78.9          
2 Afghanistan          AFG          35.253      5.9          
3 Angola               AGO          45.985     19.1          
4 Albania              ALB          12.877     57.2          
5 United Arab Emirates ARE          11.044     88.0          
6 Argentina            ARG          17.716     59.9          
  Income.Group        MupltiplyCalc
1 High income          808.2516    
2 Low income           207.9927    
3 Upper middle income  878.3135    
4 Upper middle income  736.5644    
5 High income          971.8720    
6 High income         1061.1884

Look how we now have added a new column that are multiplying our two variables. Cool huh? Try and add your own column(s) 

Let's try and remove the column again

In [24]:
data$MupltiplyCalc <- NULL
head(data)

Country.Name         Country.Code Birth.rate Internet.users
1 Aruba                ABW          10.244     78.9          
2 Afghanistan          AFG          35.253      5.9          
3 Angola               AGO          45.985     19.1          
4 Albania              ALB          12.877     57.2          
5 United Arab Emirates ARE          11.044     88.0          
6 Argentina            ARG          17.716     59.9          
  Income.Group       
1 High income        
2 Low income         
3 Upper middle income
4 Upper middle income
5 High income        
6 High income

Try and add and remove more rows from our data frame yourself. 

### Filtering a data frame 

If you want to look at,delete og model specific ROWS this is what you will be doing. Let's try and see what countries have more than 2% of internet users with the following command

In [25]:
head(data)
data$Internet.users < 2 

Country.Name         Country.Code Birth.rate Internet.users
1 Aruba                ABW          10.244     78.9          
2 Afghanistan          AFG          35.253      5.9          
3 Angola               AGO          45.985     19.1          
4 Albania              ALB          12.877     57.2          
5 United Arab Emirates ARE          11.044     88.0          
6 Argentina            ARG          17.716     59.9          
  Income.Group       
1 High income        
2 Low income         
3 Upper middle income
4 Upper middle income
5 High income        
6 High income

[1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE
 [13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [25] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [37] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [49] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE
 [61] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [73] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [85] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [97] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[109] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE
[121] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE
[133] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[145] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE
[157]  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[169] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[181] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[193] FALSE FALSE FALSE

Excersice: what countries have High income? (hint - use the same command) 

now let's try and turn this new vector into an object with all the countries that have less than 2 % internet users 

In [ ]:
filter <- data$Internet.users < 2 
data[filter,]

Let's look at the countries that have more than 40 in  birthrate and filter out these. the data frame returns the values assosiated with more than 40 in birthrate.

In [ ]:
data[data$Birth.rate > 40, ]

In [ ]:
Let us put more conditions to our filter by including countries that have more than 2 % internet users. Fill out the "?" with the correct sign to make it work: 

In [ ]:
data[data$Birth.rate > 40  & data$Internet.users ? ?,]


Try and filter out all countries with "High income". After that try and filter out Malta and inspect this countries statistics 

In [ ]:
data[data$Country.Name == "Malta",]

Try build your own function


### Q-plot and playing with visualising the data - create your own statistics.

To get qplot you need to install ggplot2 package on your R studio. If you are on jupyter notebook you don't have to do this. 

Let's try and use our dataset to do a bit of exploration with Qplot

In [ ]:
qplot(data=stats, x=Internet.users)
qplot(data=stats, x=Income.Group, y=Birth.rate, size=3, colour=I("blue"))
qplot(data=stats, x=Income.Group, y=Birth.rate, geom="boxplot")

Try and plot some more data yourself before running the next code. Try and visualise birth rate and internet users for example. try and make it the dots bigger and make a good visualisation of the phenomenon. 

In [ ]:
qplot(data=stats, x=Internet.users, y=Birth.rate, size=I(4), colour=Income.Group, size=I(5))

we can now clearly see how the income groups are located on the birth rate scale. Interesting! 

## Building your own data frames
We now know how to work with data frames but not how to create them. Our manager is urgently asking us to produce a scatterplot illustrating Birth rate and Internet usage statistics by country. This time he wants the scatterplot to be categorised by Countries' regions. He provides you with a new dataset with R vectors that he wants you to add. Ouch! let's get started. The dataset is called "CountryRegionVectors.R" and is an R-file consiting of mentioned vectors. Open the file and copy in the new vectors

In [ ]:
#Execute below code to generate three new vectors
Countries_2012_Dataset <- c("Aruba","Afghanistan","Angola","Albania","United Arab Emirates","Argentina","Armenia","Antigua and Barbuda","Australia","Austria","Azerbaijan","Burundi","Belgium","Benin","Burkina Faso","Bangladesh","Bulgaria","Bahrain","Bahamas, The","Bosnia and Herzegovina","Belarus","Belize","Bermuda","Bolivia","Brazil","Barbados","Brunei Darussalam","Bhutan","Botswana","Central African Republic","Canada","Switzerland","Chile","China","Cote d'Ivoire","Cameroon","Congo, Rep.","Colombia","Comoros","Cabo Verde","Costa Rica","Cuba","Cayman Islands","Cyprus","Czech Republic","Germany","Djibouti","Denmark","Dominican Republic","Algeria","Ecuador","Egypt, Arab Rep.","Eritrea","Spain","Estonia","Ethiopia","Finland","Fiji","France","Micronesia, Fed. Sts.","Gabon","United Kingdom","Georgia","Ghana","Guinea","Gambia, The","Guinea-Bissau","Equatorial Guinea","Greece","Grenada","Greenland","Guatemala","Guam","Guyana","Hong Kong SAR, China","Honduras","Croatia","Haiti","Hungary","Indonesia","India","Ireland","Iran, Islamic Rep.","Iraq","Iceland","Israel","Italy","Jamaica","Jordan","Japan","Kazakhstan","Kenya","Kyrgyz Republic","Cambodia","Kiribati","Korea, Rep.","Kuwait","Lao PDR","Lebanon","Liberia","Libya","St. Lucia","Liechtenstein","Sri Lanka","Lesotho","Lithuania","Luxembourg","Latvia","Macao SAR, China","Morocco","Moldova","Madagascar","Maldives","Mexico","Macedonia, FYR","Mali","Malta","Myanmar","Montenegro","Mongolia","Mozambique","Mauritania","Mauritius","Malawi","Malaysia","Namibia","New Caledonia","Niger","Nigeria","Nicaragua","Netherlands","Norway","Nepal","New Zealand","Oman","Pakistan","Panama","Peru","Philippines","Papua New Guinea","Poland","Puerto Rico","Portugal","Paraguay","French Polynesia","Qatar","Romania","Russian Federation","Rwanda","Saudi Arabia","Sudan","Senegal","Singapore","Solomon Islands","Sierra Leone","El Salvador","Somalia","Serbia","South Sudan","Sao Tome and Principe","Suriname","Slovak Republic","Slovenia","Sweden","Swaziland","Seychelles","Syrian Arab Republic","Chad","Togo","Thailand","Tajikistan","Turkmenistan","Timor-Leste","Tonga","Trinidad and Tobago","Tunisia","Turkey","Tanzania","Uganda","Ukraine","Uruguay","United States","Uzbekistan","St. Vincent and the Grenadines","Venezuela, RB","Virgin Islands (U.S.)","Vietnam","Vanuatu","West Bank and Gaza","Samoa","Yemen, Rep.","South Africa","Congo, Dem. Rep.","Zambia","Zimbabwe")
Codes_2012_Dataset <- c("ABW","AFG","AGO","ALB","ARE","ARG","ARM","ATG","AUS","AUT","AZE","BDI","BEL","BEN","BFA","BGD","BGR","BHR","BHS","BIH","BLR","BLZ","BMU","BOL","BRA","BRB","BRN","BTN","BWA","CAF","CAN","CHE","CHL","CHN","CIV","CMR","COG","COL","COM","CPV","CRI","CUB","CYM","CYP","CZE","DEU","DJI","DNK","DOM","DZA","ECU","EGY","ERI","ESP","EST","ETH","FIN","FJI","FRA","FSM","GAB","GBR","GEO","GHA","GIN","GMB","GNB","GNQ","GRC","GRD","GRL","GTM","GUM","GUY","HKG","HND","HRV","HTI","HUN","IDN","IND","IRL","IRN","IRQ","ISL","ISR","ITA","JAM","JOR","JPN","KAZ","KEN","KGZ","KHM","KIR","KOR","KWT","LAO","LBN","LBR","LBY","LCA","LIE","LKA","LSO","LTU","LUX","LVA","MAC","MAR","MDA","MDG","MDV","MEX","MKD","MLI","MLT","MMR","MNE","MNG","MOZ","MRT","MUS","MWI","MYS","NAM","NCL","NER","NGA","NIC","NLD","NOR","NPL","NZL","OMN","PAK","PAN","PER","PHL","PNG","POL","PRI","PRT","PRY","PYF","QAT","ROU","RUS","RWA","SAU","SDN","SEN","SGP","SLB","SLE","SLV","SOM","SRB","SSD","STP","SUR","SVK","SVN","SWE","SWZ","SYC","SYR","TCD","TGO","THA","TJK","TKM","TLS","TON","TTO","TUN","TUR","TZA","UGA","UKR","URY","USA","UZB","VCT","VEN","VIR","VNM","VUT","PSE","WSM","YEM","ZAF","COD","ZMB","ZWE")
Regions_2012_Dataset <- c("The Americas","Asia","Africa","Europe","Middle East","The Americas","Asia","The Americas","Oceania","Europe","Asia","Africa","Europe","Africa","Africa","Asia","Europe","Middle East","The Americas","Europe","Europe","The Americas","The Americas","The Americas","The Americas","The Americas","Asia","Asia","Africa","Africa","The Americas","Europe","The Americas","Asia","Africa","Africa","Africa","The Americas","Africa","Africa","The Americas","The Americas","The Americas","Europe","Europe","Europe","Africa","Europe","The Americas","Africa","The Americas","Africa","Africa","Europe","Europe","Africa","Europe","Oceania","Europe","Oceania","Africa","Europe","Asia","Africa","Africa","Africa","Africa","Africa","Europe","The Americas","The Americas","The Americas","Oceania","The Americas","Asia","The Americas","Europe","The Americas","Europe","Asia","Asia","Europe","Middle East","Middle East","Europe","Middle East","Europe","The Americas","Middle East","Asia","Asia","Africa","Asia","Asia","Oceania","Asia","Middle East","Asia","Middle East","Africa","Africa","The Americas","Europe","Asia","Africa","Europe","Europe","Europe","Asia","Africa","Europe","Africa","Asia","The Americas","Europe","Africa","Europe","Asia","Europe","Asia","Africa","Africa","Africa","Africa","Asia","Africa","Oceania","Africa","Africa","The Americas","Europe","Europe","Asia","Oceania","Middle East","Asia","The Americas","The Americas","Asia","Oceania","Europe","The Americas","Europe","The Americas","Oceania","Middle East","Europe","Europe","Africa","Middle East","Africa","Africa","Asia","Oceania","Africa","The Americas","Africa","Europe","Africa","Africa","The Americas","Europe","Europe","Europe","Africa","Africa","Middle East","Africa","Africa","Asia","Asia","Asia","Asia","Oceania","The Americas","Africa","Europe","Africa","Africa","Europe","The Americas","The Americas","Asia","The Americas","The Americas","The Americas","Asia","Oceania","Middle East","Oceania","Middle East","Africa","Africa","Africa","Africa")

Let's put these into a dataframe and check the head function again to see the top rows 

In [ ]:
mydf <- data.frame(Countries_2012_Dataset, Codes_2012_Dataset, Regions_2012_Dataset)
head(mydf)

Let's try and change the names of each column with the function colnames(mydf) &lt;- c("Country", "Code", "Region") 

In [ ]:
colnames(mydf) <- c("Country", "Code", "Region") 
head(mydf)

Let's try and remove and recreate this dataframe with names right away and save ourselves some code

In [ ]:
rm(mydf)
mydf <- data.frame(Country=Countries_2012_Dataset, Code=Codes_2012_Dataset, Regions= Regions_2012_Dataset)
head(mydf)

Excercise: Try and name the columns the way you want.

## Lets now try to MERGE dataframes together. 
We now have two dataframes. The mydf and the stats dataframe. Let's try and merge them

start by exploring both datasets to make sure they are there

In [ ]:
head(stats)
head(mydf)

In [ ]:
## now we will merge the two and tell R to match by country
merged <- merge(stats, mydf, by.x ="Country.Code", by.y = "Code" )
head(merged)

#we now have a dublicate colulmn in country.name and Country. Let's apply NULL to it

merged$Country <- NULL 
str(merged)

# Last excercise
Now try and do the job the bank boss asked us to do with the new dataframe! 